jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java

```diff
@@ -0,0 +1,735 @@
+     * 
+     * @author Hannes Wellmann
+     * @param <V> the type of vertices in the imported graph
+     */
+    public static class TSPLIBFileData<V>
+    {
+
+        private String name;
+        private String type;
+        private String comment;
+        private Integer dimension;
+        private String edgeWeightType;
+        private String nodeCoordType;
+        private boolean hasDistinctLocations;
+        private Boolean hasDistinctEdges;
+
+        private Graph<V, DefaultWeightedEdge> graph;
+        private List<Integer> tour;
+
+        // getters for read public data
+
+        /**
+         * Returns the value of the <em>NAME</em> keyword in the imported file.
+         * 
+         * @return the value of the <em>NAME</em> keyword
+         */
+        public String getName()
+        {
+            return name;
+        }
+
+        /**
+         * Returns the value of the <em>TYPE</em> keyword in the imported file.
+         * 
+         * @return the value of the <em>TYPE</em> keyword
+         */
+        public String getType()
+        {
+            return type;
+        }
+
+        /**
+         * Returns the joined value of the <em>COMMENT</em> keyword in the imported file.
+         * 
+         * @return the value of the <em>COMMENT</em> keyword
+         */
+        public String getComment()
+        {
+            return comment;
+        }
+
+        /**
+         * Returns the value of the <em>DIMENSION</em> keyword in the imported file.
+         * 
+         * @return the value of the <em>DIMENSION</em> keyword
+         */
+        public Integer getDimension()
+        {
+            return dimension;
+        }
+
+        /**
+         * Returns the value of the <em>EDGE_WEIGHT_TYPE</em> keyword in the imported file.
+         * 
+         * @return the value of the <em>EDGE_WEIGHT_TYPE</em> keyword
+         */
+        public String getEdgeWeightType()
+        {
+            return edgeWeightType;
+        }
+
+        /**
+         * Returns the value of the <em>NODE_COORD_TYPE</em> keyword in the imported file.
+         * 
+         * @return the value of the <em>NODE_COORD_TYPE</em> keyword
+         */
+        public String getNodeCoordType()
+        {
+            return nodeCoordType;
+        }
+
+        /**
+         * Returns the complete {@link Graph} of location connections built from the imported file
+         * or null if no graph was imported.
+         * 
+         * @return the complete {@code Graph} of location connections
+         */
+        public Graph<V, DefaultWeightedEdge> getGraph()
+        {
+            return graph;
+        }
+
+        /**
+         * Returns the {@link List} of {@link Integer Integers} describing vertex order of the tour
+         * from the imported file or null if no tour was imported.
+         * 
+         * @return the tour described by the list of vertex indices
+         */
+        public List<Integer> getTour()
```

Comment:
What does it mean to return a list of integers `List<Integer>`?  Vertices in JGraphT may be any type of object, you need to translate the integers to vertices.

Reply:
The problem is that a TSPLIB-file containing a tour does not need to contain the node coordinates defining the TSP-problem instance.
In the cases I know there is a file containing the `NODE_COORD_SECTION` with the coordinates of all nodes (where each node also has a number) and a separate file that contains for example the optimal tour or the best known tour. That tour is described by a List of node-numbers that correspond to the numbers in the file with the `NODE_COORD_SECTION`. It is possible but definitely not always the case that the node coordinates and the tour is contained in one file.

I agree that a List of vertices is better. I thought about multiple options to achieve this. The one I prefer is the following:

- Let `getTour()` return a `List<V>` that is only not null, if a file contains a `TOUR_SECTION`.
- create a separate method `importTour(List<V>)` that has a List of vertices as their parameter that is ordered according to the node numbers and returns the List of vertices in the order of the tour. This enables import from a file that only contains a `TOUR_SECTION` . I think this method should create a MetaData object too (but maybe this has to be discussed).
- To enable the latter method I would also add a `getOrderedVertices()` method to the MetaData class that returns the vertices imported from a file with `NODE_COORD_SECTION` in their "number"-order.